datafusion/core/src/physical_plan/topk/mod.rs

```diff
@@ -0,0 +1,516 @@
+                // update memory accounting
+                let prev_size = owned_row_size(&prev_min.row);
+                assert!(self.owned_row_bytes >= prev_size);
+                self.owned_row_bytes -= prev_size;
+            }
+        }
+    }
+
+    /// Returns the values stored in this heap, from values low to high, as a single
+    /// [`RecordBatch`]
+    pub fn emit(&self, schema: SchemaRef) -> Result<RecordBatch> {
+        // Indicies for each row within its respective RecordBatch
+        let indicies: Vec<_> = self
+            .inner
+            .iter()
+            .enumerate()
+            .map(|(i, k)| (i, k.index))
+            .collect();
+
+        let num_columns = {
+            let Some(first_value) = self.inner.get(0) else {
+                return Ok(RecordBatch::new_empty(schema));
+            };
+            self.store
+                .get(first_value.batch_id)
+                .expect("invalid batch id")
+                .batch
+                .num_columns()
+        };
+
+        // build the output columns one at time, using the
+        // `interleave` kernel to pick rows from different arrays
+        let output_columns: Vec<_> = (0..num_columns)
+            .map(|col| {
+                let input_arrays: Vec<_> = self
+                    .inner
+                    .iter()
+                    .map(|k| {
+                        let entry =
+                            self.store.get(k.batch_id).expect("invalid stored batch id");
+                        entry.batch.column(col) as &dyn Array
+                    })
+                    .collect();
+
+                // at this point `indices` contains indexes within the
+                // rows and `input_arrays` contains a reference to the
+                // relevant Array for that index. `interleave` pulls
+                // them together into a single new array
+                Ok(interleave(&input_arrays, &indicies)?)
+            })
+            .collect::<Result<_>>()?;
+
+        Ok(RecordBatch::try_new(schema, output_columns)?)
+    }
+
+    /// return the size of memory used by this heap, in bytes
+    fn size(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + (self.inner.capacity() * std::mem::size_of::<TopKRow>())
+            + self.store.size()
+            + self.owned_row_bytes
+    }
+}
+
+/// Size of memory owned by `row` until row::size() is available
+/// TODO file upstream ticket in arrow-rs to add this
+fn owned_row_size(row: &OwnedRow) -> usize {
+    std::mem::size_of_val(row) + row.as_ref().len() // underlying data, doesn't account for capacity
+}
+
+/// Represents one of the top K rows. Orders according to `OwnedRow`
+#[derive(Debug, PartialEq)]
+struct TopKRow {
+    /// the value of the sort key for this row
+    row: OwnedRow,
+    /// the index in this record batch the row came from
+    index: usize,
+    /// the RecordBatch this row came from: an id into a [`RecordBatchStore`]
+    batch_id: u32,
```

Comment:
It occurs to me that you could potentially just store `Arc<RecordBatch>` here, potentially avoiding the need for `RecordBatchStore`.

Memory accounting would be a bit more complex, but could probably make use of [Arc::into_inner](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_inner) to detect if the last reference

Reply:
Yeah, the memory accounting was what I was worried about -- maybe I could use a HashSet and compare the underlying pointers or something 🤔 

But for what it is worth, when profiled this most of the time seems to have gone into managing `Row`s 

![Screenshot 2023-08-11 at 1 04 40 PM](https://github.com/apache/arrow-datafusion/assets/490673/96067201-0159-415e-b60c-503ee93c4a1e)

(which FWIW I plan to file a suggestion in arrow-rs shortly about how to avoid the Row allocations by reusing existing `OwnedRow`)

Reply:
You could also just use `Vec<u8>`, the only reason to use OwnedRow is if you want to be able to go back, which given we retain the source RecordBatch seems unlikely to be necessary.

Reply:
`Vec<u8>` would be fine -- the owned row is simply used for comparison. As you say this PR doesn't use it to go back to Arrays. I'll give it a try